subsys/net/lib/prometheus/formatter.c

```diff
@@ -0,0 +1,226 @@
+/*
+ * Copyright (c) 2024 Mustafa Abdullah Kus, Sparse Technology
+ *
+ * SPDX-License-Identifier: Apache-2.0
+ */
+
+#include <zephyr/net/prometheus/formatter.h>
+#include <zephyr/net/prometheus/collector.h>
+#include <zephyr/net/prometheus/metric.h>
+#include <zephyr/net/prometheus/histogram.h>
+#include <zephyr/net/prometheus/summary.h>
+#include <zephyr/net/prometheus/gauge.h>
+#include <zephyr/net/prometheus/counter.h>
+
+#include <stdarg.h>
+#include <stdio.h>
+#include <string.h>
+#include <stddef.h>
+
+#include <zephyr/kernel.h>
+
+#include <zephyr/logging/log.h>
+LOG_MODULE_REGISTER(pm_formatter, CONFIG_PROMETHEUS_LOG_LEVEL);
+
+static int write_to_buffer(char *buffer, size_t buffer_size, const char *str)
+{
+	/* helper function to write formatted string to buffer, handling buffer size */
+	int len = snprintf(buffer + strlen(buffer), buffer_size - strlen(buffer), "%s", str);
+
+	if (len >= 0 && len < buffer_size - strlen(buffer)) {
+		return 1;
+	}
+
+	/* error (insufficient buffer size) */
+	return 0;
+}
+
+static int write_metric_to_buffer(char *buffer, size_t buffer_size, const char *format, ...)
+{
+	/* helper function to write formatted metric to buffer */
+	char temp_buffer[CONFIG_PROMETHEUS_METRIC_FORMAT_BUFFER_SIZE];
+	va_list args;
+
+	va_start(args, format);
+	vsnprintf(temp_buffer, sizeof(temp_buffer), format, args);
+	va_end(args);
+
+	return write_to_buffer(buffer, buffer_size, temp_buffer);
+}
+
+void prometheus_format_exposition(const struct prometheus_collector *collector, char *buffer,
+				  size_t buffer_size)
+{
+	int written = 0;
+
+	if (collector == NULL || buffer == NULL || buffer_size == 0) {
+		LOG_ERR("Invalid arguments");
+		return;
+	}
+
+	/* iterate through each metric in the collector */
+	for (size_t ind = 0; ind < collector->size; ind++) {
+
+		const struct prometheus_metric *metric = collector->metric[ind];
+
+		/* write HELP line if available */
+		if (metric->description[0] != '\0') {
+			if (!write_metric_to_buffer(buffer + written, buffer_size - written,
+						    "# HELP %s %s\n", metric->name,
+						    metric->description)) {
+				LOG_WRN("Error writing to buffer");
```

Comment:
Are these errors or warnings, the string indicates an error so should we use `LOG_ERR` here and in other warning prints below.